pisek/tools/judgelib/judge-shuffle.cc

```diff
@@ -0,0 +1,350 @@
+	if (read_pos >= pg->end) {
+		this_page = pg = pg->next;
+		if (!pg)
+			return NULL;
+		read_pos = pg->buf;
+	}
+	return read_pos;
+}
+
+/*** Reading and shuffling ***/
+
+struct tok {
+	const char *token;
+	uint hash;
+
+	void compute_hash()
+	{
+		const char *x = token;
+		hash = 1;
+		while (*x)
+			hash = (hash * 0x6011) + *x++;
+	}
+
+	int compare(const tok with) const
+	{
+		if (hash < with.hash)
+			return -1;
+		if (hash > with.hash)
+			return 1;
+		return strcmp(token, with.token);
+	}
+
+	bool operator < (const tok &with) const
+	{
+		return compare(with) < 0;
+	}
+
+	bool operator != (const tok &with) const
+	{
+		return compare(with) != 0;
+	}
+};
+
+struct line {
+	tok *toks;
+	uint len;
+	uint hash;
+	uint orig_line;
+
+	void compute_hash()
+	{
+		hash = 1;
+		for (uint i=0; i < len; i++)
+			hash = (hash * 0x6011) + toks[i].hash;
+	}
+
+	int compare(const line &with) const
+	{
+		if (hash < with.hash)
+			return -1;
+		if (hash > with.hash)
+			return 1;
+
+		if (len < with.len)
+			return -1;
+		if (len > with.len)
+			return 1;
+
+		for (uint i=0; i < len; i++) {
+			int c = toks[i].compare(with.toks[i]);
+			if (c)
+				return c;
+		}
+		return 0;
+	}
+
+	bool operator < (const line &with) const
+	{
+		return compare(with) < 0;
+	}
+
+	bool operator != (const line &with) const
+	{
+		return compare(with) != 0;
+	}
+};
+
+struct shuffler {
+	tokbuf *tb;
+	vector<tok> toks;
+	vector<line> lines;
+	uint num_lines;
+
+	void read(tokenizer *tizer);
+	void slurp(tokenizer *tizer);
+};
+
+void shuffler::slurp(tokenizer *tizer)
+{
+	tb = new tokbuf;
+	char *token;
+	bool nl = true;
+
+	while (token = tizer->get_token()) {
+		if (token[0]) {
+			nl = false;
+			if (ignore_case)
+				for (char *c=token; *c; c++)
+					if (*c >= 'a' && *c <= 'z')
+						*c = *c - 'a' + 'A';
+		} else if (!ignore_nl) {
+			if (nl && ignore_empty)
+				continue;
+			nl = true;
+		}
+		tb->add_token(token, tizer->toksize);
+	}
+
+	if (!nl)
+		tb->add_token("", 0);
+	tb->close();
+}
+
+void shuffler::read(tokenizer *tizer)
+{
+	slurp(tizer);
+	toks.resize(tb->num_tokens);
+	lines.resize(tb->num_lines + 1);
+
+	tok *t = toks.data();
+	line *l = lines.data();
+	l->toks = t;
+	num_lines = 0;
+	for (const char *x = tb->get_first(); x; x = tb->get_next()) {
+		if (*x) {
+			t->token = x;
+			t->compute_hash();
+			t++;
+		} else {
+			l->len = t - l->toks;
+			if (shuffle_words) {
+				auto beg = toks.begin() + (l->toks - toks.data());
+				sort(beg, beg + l->len);
+			}
+			l->compute_hash();
+			l->orig_line = ++num_lines;
+			l++;
+			l->toks = t;
+		}
+	}
+	assert(l->toks == t);
+	assert(num_lines == tb->num_lines);
+
+	if (shuffle_lines)
+		sort(lines.begin(), lines.end());
+
+#if 0
+	for (uint i=0; i < num_lines; i++) {
+		line *l = &lines[i];
+		printf("%08x %5d: ", l->hash, l->orig_line);
+		for (uint j=0; j < l->len; j++) {
+			tok *t = &l->toks[j];
+			printf("<%08x|%s>", t->hash, t->token);
+		}
+		printf("\n");
+	}
+#endif
```

Comment:
Nit: Do we want to leave in debugging code?

Reply:
Not having different version then upstream is higher priority.